### ISSUE TITLE
[ObjectLiteral] class is identified as type "String" when deserialised if TypeNameHandling enabled

### ISSUE THREAD

**Issue by ProductiveRage** [bridgedotnet/Bridge.Newtonsoft.Json#53]
When an `[ObjectLiteral]` class is serialised and then deserialised and `TypeNameHandling` is set to `Objects` then the `$type` of the deserialised object gets corrupted and is returned as string.

To illustrate:

```csharp
public class Program
{
    public static void Main()
    {
        var settings = new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Objects };
        var x = new Person("Test");
        Console.WriteLine("Source is of type " + x.GetType());

        var json = JsonConvert.SerializeObject(x, settings);
        Console.WriteLine(json);

        var clone = JsonConvert.DeserializeObject<Person>(json, settings);
        Console.WriteLine("Clone is of type " + clone.GetType());
    }
}

[ObjectLiteral(ObjectCreateMode.Constructor)]
public class Person
{
    public Person(string name)
    {
        Name = name;
    }
    public string Name { get; }
}
```

Available at https://deck.net/efb481ec2abe2a29477f394397d78a69

The expected output is:

```
Source is of type Demo.Person
{"$type":"Demo.Person, Demo","Name":"Test"}
Clone is of type Demo.Person
```

but the actual output is: 

```
Source is of type Demo.Person
{"$type":"Demo.Person, Demo","Name":"Test"}
Clone is of type String
```

I'm not sure if the problem lies more with Bridge or with the Json.NET implementation. I think that one way to fix it would be to change Bridge's `getType` implementation because it currently checks for a `$type` property and returns that, presuming that it will be a **Type** instance - but, of course, it's a string when it comes from JSON of this format. So one option may be to change this bit from Bridge's `getType`.

```js
if (instance.$type) {
    return instance.$type;
}
```

to this:

```js
if (instance.$type) {
    if (Bridge.isString(instance.$type)) {
        instance.$type = Bridge.Reflection.getType(instance.$type);
    }

    return instance.$type;
}
```

That would fix it when `TypeNameHandling` is enabled but there is a different problem when it's not. Without `TypeNameHandling`, the code above works correctly. _However,_ if it's changed so that there is a nested type reference in the `[ObjectLiteral]` class then it starts to go wrong -

```csharp
public class Program
{
	public static void Main()
	{
		var settings = new JsonSerializerSettings { };
		var x = new Person("Test", new Person("Boss", null));
		Console.WriteLine("Source is of type " + x.GetType());
		Console.WriteLine("Source.Manager is of type " + x.Manager.GetType());

		var json = JsonConvert.SerializeObject(x, settings);
		Console.WriteLine(json);

		var clone = JsonConvert.DeserializeObject<Person>(json, settings);
		Console.WriteLine("Clone is of type " + clone.GetType());
		Console.WriteLine("Clone.Manager is of type " + clone.Manager.GetType());
	}
}

[ObjectLiteral(ObjectCreateMode.Constructor)]
public class Person
{
	public Person(string name, Person manager)
	{
		Name = name;
		Manager = manager;
	}
	public string Name { get; }
	public Person Manager { get; }
}
```

Expected output is

```
Source is of type Demo.Person
Source.Manager is of type Demo.Person
{"Manager":{"Manager":null,"Name":"Boss"},"Name":"Test"}
Clone is of type Demo.Person
Clone.Manager is of type Demo.Person
```

but the actual output is:

```
Source is of type Demo.Person
Source.Manager is of type Demo.Person
{"Manager":{"Manager":null,"Name":"Boss"},"Name":"Test"}
Clone is of type Demo.Person
Clone.Manager is of type System.Object
```

(The type name is correct on the "clone" but not on its "Manager" property).

Since there is no `$type` property set on all nested references when `TypeNameHandling` is disabled, maybe a separate pass is required on the output of the deserialisation process to fix things like this up? For cases where `TypeNameHandling` is enabled, the fix-up step would translate those `$type` strings into type instances. For cases where `TypeNameHandling` is not enabled, a `$type` property could be set by using reflection metadata (we know that the **Manager** property is of type **Person** and so we could set the `$type` property to the **Person** type - this wouldn't handle inheritance correctly but I think that that's an accepted limitation of serialising without `TypeNameHandling`).

This "fix-up step" could potentially also address the **DateTime** deserialisation problem that `[ObjectLiteral]` classes have that I documented in Issue #52. Having the fix-up step performed by this library should mean that no changes to the core Bridge library are required.

I will have some time this week to look into this, if that's any help - I don't know if you will find it easier (in terms of time / resources / etc..) to fix it internally or to have to review pull requests?

We're keen to get this working with `[ObjectLiteral]` classes because the deserialisation process is much much faster than otherwise - I've been investigating performance last week and this week, hence all the bug reports! I think that I only have one more to raise..

**Comment by vladsch** [comment]
Related with #52, so the fix from #52 should affect on #53

**Comment by fabriciomurta** [comment]
Hi!

Just run your test case with `dev` branch.
- create new ClassLibrary project on .NET Framework 4
- install Bridge.Newtonsoft.Json 1.3.0 package (bridge 16.5.0 comes in the bundle)
- Paste the code above within the `namespace` block of the `Class1.cs` file (for simplicity)
- Build
- Open `bin/Debug/Bridge/index.html` in google chrome
- Open console (F12)
Result:
![image](https://user-images.githubusercontent.com/10730372/32421241-768e8290-c27d-11e7-9ec9-4781482055cf.png)

That's the correct result, right?
